### DISCUSSION TITLE
How to restrict access to the client-secret API so that a client can only GET/POST on it's own secret?

### DISCUSSION THREAD

**Discussion by bhayden53** [keycloak/keycloak#34317] [Q&A]
Hello,

I am working on an application which includes keycloak, and a set of multiple service account clients representing different applications which need access to external resources. Those applications will access those resources via tokens retrieved using their client_secret. All standard stuff.

Our institution requires secret rotation to be automated and periodic, so I need to provide these clients with programmatic access to the client-secret endpoint. The problem is that I can't seem to find a way to restrict access so that a client can only manage it's own secret. The realm-level view-clients/manage-clients roles allow full access to all clients in the realm. 

I have tried removing those roles and setting up fine-grained permissions, but as far as I can tell the client-secret endpoint requires the realm-level roles. It's possible I'm setting up the fine-grained permissions incorrectly, but I think I've gotten it right.

Is it possible to do what I'm trying to do using the standard keycloak client-secret endpoint? Or do I need a custom keycloak or some other intermediary layer?

**Comment by darius-m** [comment]
From the documentation, the Signed JWT token or a X509 Certificate may make more sense if you plan on rotating the secrets automatically. The Signed JWT (https://www.keycloak.org/docs/latest/server_admin/index.html#_client-credentials) seems to be designed specifically for this use-case, using the "Use JWKS URL" option.

**Reply by darius-m** [reply]
Also, have you tried the instructions in this section? https://www.keycloak.org/docs/latest/server_admin/index.html#_secret_rotation

**Reply by bhayden53** [reply]
> Also, have you tried the instructions in this section? https://www.keycloak.org/docs/latest/server_admin/index.html#_secret_rotation

This feature is not released out of development yet, and it doesn't solve the problem of all clients in the realm having access to all other client's keys (due to needing the view-clients and manage-clients roles to access even their own secret).

I will have to study up on the JWKS option; for some paths in my application, I am using oauth2-proxy and traefik with a middleware auth flow, and it's not clear off the top of my head what I would need to modify on that side to get something other than a client id/secret flow working.

**Reply by darius-m** [reply]
I see... I have not looked into this use-case closely, so I can't help you much further than this. For the scenario you have described you may indeed need to use a plugin that adds additional constraints on the requests, as I don't think the functionality is available out of the box.

**Reply by bhayden53** [reply]
From my initial look, it seems like the client still needs to know their secret in the JWKS model, it's just the authorizing application that will use the JWKS URL? In which case the client would still need to be able to rotate their individual secret. I'll keep looking at it though.

Thanks for the responses!

**Reply by darius-m** [reply]
If I understand it correctly, JWKS is a mechanism that allows signing the client requests with a private key (and Keycloak is able to verify them using the associated public key), instead of sending a client ID + secret for authentication. The JWKS URL is an endpoint that allows Keycloak to automatically fetch the client's public keys from an URL controlled by the client, so the secret rotation takes place on the client's end, instead of the SSO.
